napari/components/layerlist.py

```diff
@@ -98,23 +96,14 @@ def move_selected(self, index, insert):
         insert : int
             Index that item(s) will be inserted at
         """
-        total = len(self)
-        indices = list(range(total))
         if not self[index].selected:
             self.unselect_all()
             self[index].selected = True
-        selected = [i for i in range(total) if self[i].selected]
-
-        # remove all indices to be moved
-        for i in selected:
-            indices.remove(i)
-        # adjust offset based on selected indices to move
-        offset = sum([i < insert and i != index for i in selected])
-        # insert indices to be moved at correct start
-        for insert_idx, elem_idx in enumerate(selected, start=insert - offset):
-            indices.insert(insert_idx, elem_idx)
-        # reorder list
-        self.move_multiple(indices, 0)
+            moving = (index,)
+        else:
+            moving = [i for i, item in enumerate(self) if item.selected]
+        offset = insert >= index
```

Comment:
I had to add this offset in here, but I think it might actually belong in `move_multiple` @tlambert03 - look at `viewer.layers.move_multiple((0, 1), 2)` I think it doesn't do the right thing, it's off by one, but it's late now so I might be wrong about this!!! Will look more tomorrow

Reply:
we can certainly change this, but currently `move_multiple((0, 1), 2)` should be a no-op... it's saying "insert the items at index 0 and 1 right before the item at index 2" (And they already are just before index 2)... just like `a.insert(2, 'something')` is saying "insert 'something' right before the item at index 2".  
```python
a = EventedList(range(10))
a.move_multiple((0, 1), 0)
a.move_multiple((0, 1), 1)
a.move_multiple((0, 1), 2)
a == list(range(10)  # True ... hasn't changed
```
thoughts?

Reply:
> it's saying "insert the items at index 0 and 1 right before the item at index 2" (And they already are just before index 2)... just like a.insert(2, 'something') is saying "insert 'something' right before the item at index 2".

Interesting, if we just think about the `a.move` case I would have expected `a.move(old_index, new_index)` to have meant that `a[new_index]` after the move has the same value as `a[old_index]` before the move, but this is not generally the case. Note that for the `a.insert(index, 'something')` example it is always the case that `a[index]='something'` after the insertion. I would not have expected `a.move(4, 5)` to be a no-op, but to move the item at 4 to be at 5.


Similarly, for example if
```python
a = TypedEventedList(['a', 'b', 'c', 'd', 'e', 'f', 'g'])
a[1] # `b`
a.move(1, 3)
a[3] # `d` (I would have expected `b`, but that is at a[2]) 
```
but note though the following
```python
a = TypedEventedList(['a', 'b', 'c', 'd', 'e', 'f', 'g'])
a[5] # `f`
a.move(5, 3)
a[5] # `f` (which is what I would expected)
```
The difference here is something is moving to a lower index. I think for the `a.move` case you do need an `offset = old_index < index`.

If the `a.move` behaves this way, then it's natural for `a.move_multiple` to behave similarly, I guess the `a.move_multiple((old_index,), new_index)` should always behave the same as `a.move(old_index, new_index)` and then the question is for `a.move_multiple((old_index1, old_index2), new_index)` with `old_index1 < old_index2` I guess one has to make a choice and I think it's fine to say after the move `a[new_index]` should have the value of `a[old_index2]` before the move.

In your example about then
```python
a = EventedList(range(10))
a.move_multiple((0, 1), 0)
a.move_multiple((0, 1), 1)
a == list(range(10)  # True ... hasn't changed
a.move_multiple((0, 1), 2)
a == [2, 0, 1] + range(3, 10) # Now the list has changed and a[2] == 1
```

I'll note that all of this was just how I would have guessed it was working, so if there is compelling prior art or other reasons to do it with the "insert right before the item at index 2" that's an option too, but curious how you find my reasoning.

Reply:
I agree, it's weird at first... and I also wanted those simple cases to have the `dest_index` be in post-move space, and I understand it flips if you're moving up vs moving back ... but I went with having the user provide "pre-move" coordinates instead of "post-move" coordinates following patterns I saw elsewhere (such as Qt's [MoveRows/BeginMoveRows](https://doc.qt.io/qt-5/qabstractitemmodel.html#beginMoveRows)), and here's why I was ultimately swayed:

first, I totally agree that `move_multiple((old,), new)` has to equal `move(old, new)`... and since move_multiple is much harder, let's take that case.  It's simple if you think about adjacent objects, like `move_multiple((0,1), 2)` ... but it's much harder if you consider that the moving objects needn't be adjacent. 
```python
a = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
a.move_multiple((2, 3, 7, 9), 5)
```
If we ask the user to provide pre-move coordinates, they look at the current list and say "i want the 2,3,7 & 9 to move together to right before the 5".  That's very much like what a drag and drop would do, where you have a handful of collected/dragged pre-move indices, and then a specific position (in pre move space) where they release the mouse and drop things.  so this hooks up to that use case nicely.
If we ask the user to provide post-move coordinates... then _they_ have to do all that math and say "ok... I'm moving the 2 and the 3 back, but the 7 and the 9 up ... so that means that after the move I guess I want them to be at ...", and then they're going to have to change that logic depending on whether part of the selection is before the destination or not.

So providing pre-move coordinates sometimes yields the undesirable effect that `dest_index` is not always where the objects land after the move, but using post-move coordinates requires being careful whenever you call the function to make sure that you call it just right depending on exactly what you're moving, making the API a bit more challenging to use.

Reply:
I'll also point out that if you want to know where they land after the move, the `moved` event has an `event.new_index` value

Reply:
and actually, currently the `move_multiple` function returns the number of successful moves (which is probably always just going to be the number of objects).  If we stick with the current pre-move space input, maybe we should have the function return the destination index in post-move space?

Reply:
> If we ask the user to provide post-move coordinates... then they have to do all that math and say "ok... I'm moving the 2 and the 3 back, but the 7 and the 9 up ... so that means that after the move I guess I want them to be at ...", and then they're going to have to change that logic depending on whether part of the selection is before the destination or not.

I find this compelling

> and actually, currently the move_multiple function returns the number of successful moves (which is probably always just going to be the number of objects). If we stick with the current pre-move space input, maybe we should have the function return the destination index in post-move space?

yes this might be nicer, but i don't really have a strong feeling on this one.

ok - this all checks out, I see the reasoning for it. As long as it's all clearly documented I think users will figure it out pretty quickly. Definitely @jni should give this stuff a play before it all merges, and I'm curious how he'll view this, but I think this is as fine a choice as any.

Reply:
cool.  I'm also realizing that we really don't need both `move()` and `move_multiple()` ... rather we can just check if the first argument is iterable or not and use the move_multiple logic if so... yeah?

Reply:
> I'm also realizing that we really don't need both move() and move_multiple() ... rather we can just check if the first argument is iterable or not and use the move_multiple logic if so... yeah?

Yes! wow - definitely just having move would be amazing!! I also think if second arg not provided then it defaults to `0`!! So clean!!!

Reply:
just rereading this thread now @tlambert03 - it sounds like this comment https://github.com/napari/napari/pull/1504/files#r460995503 was a good idea, but hasn't been implemented yet. Nothing to hold up this PR for, but wanted to flag so we don't forget the good idea. I can make an issue too

Reply:
issue made https://github.com/napari/napari/issues/1845